File: examples/canonical-analytics/go/internal/analytics/events.go
package analytics

import (
	"sort"

	"analytics-workbench-go/internal/data"
	"analytics-workbench-go/internal/filters"
)

type EventCountSeries struct {
	Dates         []string
	Counts        []int
	ByEnvironment map[string][]int
}

func AggregateEvents(events []data.Event, f filters.FilterState) EventCountSeries {
	if len(events) == 0 {
		return EventCountSeries{ByEnvironment: make(map[string][]int)}
	}

	dateSet := make(map[string]bool)
	envSet := make(map[string]bool)
	countsByDateEnv := make(map[string]map[string]int) // date -> env -> count
	totalByDate := make(map[string]int)

	for _, e := range events {
		if !f.ContainsEnvironment(e.Environment) {
			continue
		}
		
		t := parseDate(e.Timestamp)
		if !f.InDateRange(t) {
			continue
		}
		if !f.ContainsService(e.Service) {
			continue
		}

		dateStr := e.Timestamp[:10]
		dateSet[dateStr] = true
		envSet[e.Environment] = true

		if countsByDateEnv[dateStr] == nil {
			countsByDateEnv[dateStr] = make(map[string]int)
		}
		countsByDateEnv[dateStr][e.Environment] += e.Count
		totalByDate[dateStr] += e.Count
	}

	if len(dateSet) == 0 {
		return EventCountSeries{ByEnvironment: make(map[string][]int)}
	}

	var dates []string
	for d := range dateSet {
		dates = append(dates, d)
	}
	sort.Strings(dates)

	counts := make([]int, len(dates))
	byEnv := make(map[string][]int)
	for env := range envSet {
		byEnv[env] = make([]int, len(dates))
	}

	for i, d := range dates {
		counts[i] = totalByDate[d]
		for env := range envSet {
			byEnv[env][i] = countsByDateEnv[d][env]
		}
	}

	return EventCountSeries{
		Dates:         dates,
		Counts:        counts,
		ByEnvironment: byEnv,
	}
}

File: examples/canonical-analytics/rust/src/analytics/events.rs
use super::{match_date, match_multi, EventCountSeries};
use crate::data::models::Event;
use crate::filters::FilterState;
use std::collections::{BTreeMap, HashMap};

pub fn aggregate_events(data: &[Event], filters: &FilterState) -> EventCountSeries {
    if data.is_empty() {
        return EventCountSeries::default();
    }

    let mut env_dates: HashMap<String, BTreeMap<String, i64>> = HashMap::new();
    let mut all_dates: BTreeMap<String, i64> = BTreeMap::new();
    let mut environments_seen = Vec::new();

    for ev in data {
        if !match_date(&ev.timestamp, &filters.date_from, &filters.date_to) {
            continue;
        }
        if !match_multi(&ev.service, &filters.services) {
            continue;
        }
        if !match_multi(&ev.environment, &filters.environment) {
            continue;
        }

        let date = ev.timestamp.split('T').next().unwrap_or(&ev.timestamp).to_string();
        
        *all_dates.entry(date.clone()).or_insert(0) += ev.count;

        let env_map = env_dates.entry(ev.environment.clone()).or_default();
        *env_map.entry(date).or_insert(0) += ev.count;

        if !environments_seen.contains(&ev.environment) {
            environments_seen.push(ev.environment.clone());
        }
    }

    let dates: Vec<String> = all_dates.keys().cloned().collect();
    let counts: Vec<i64> = all_dates.values().cloned().collect();

    let mut by_environment = HashMap::new();
    for env in environments_seen {
        let env_map = env_dates.get(&env).unwrap();
        let env_counts: Vec<i64> = dates.iter().map(|d| *env_map.get(d).unwrap_or(&0)).collect();
        by_environment.insert(env, env_counts);
    }

    EventCountSeries {
        dates,
        counts,
        by_environment,
    }
}
